assembly/src/assembler/module_provider.rs

```diff
@@ -0,0 +1,59 @@
+use super::{BTreeMap, Library, LibraryError, Module, ProcedureId, Vec};
+
+// MODULE PROVIDER
+// ================================================================================================
+
+/// A module provider that will store the procedure ASTs used during the compilation process.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct ModuleProvider {
```

Comment:
A comment from the previous review: I wonder if we should track a set of library root namespaces added to the provider. This way, if two libraries with the same root namespace get added two or more times, we can give a more informative error.

Reply:
To give further context, we are tracking that but on a module level (i.e. we don't allow duplicated modules).

This will allow the user to have modules defined in different libraries for the same namespace, but will prevent users from duplicating modules. It is ofc trivial to prevent users from using the same namespace in different libraries, so it is mostly a design decision.

The advantage of tracking per module is that we don't have to bundle modules before compilation - we can just use individual module definitions and add them as they are built (i.e. multiple source files). If we opt to track per namespace, we need to assemble these files into a single bundle, and only then add this to the library.

Reply:
Issue added #582